# **Assignment 2**

## **ECON8502 - Structural Microeconometrics**

### *Conor Bayliss*

#### **Setup**

The first thing we do is run the setup code. This reads in the data and creates a NamedTuple "data_mle".

In [1]:
using CSV, DataFrames, DataFramesMeta, Optim, ForwardDiff
include("C:\\Users\\bayle\\Documents\\Github\\micro_labour\\Transfers.jl")

data = @chain begin
    CSV.read("C:\\Users\\bayle\\Documents\\Github\\micro_labour\\hw1\\hw1_micro_labour_data\\MainPanelFile.csv", DataFrame, missingstring = "NA")
    #@select :MID :year :wage :hrs :earn :SOI :CPIU :WelfH :FSInd :num_child :age
    @subset :year.>=1985 :year.<=2010 # 1985-2010 subset
    @transform :AFDC = :WelfH.>0 # AFDC dummy
    @rename :FS = :FSInd # FS dummy
    @transform :P = :FS + :AFDC :H = min.(2, round.(Union{Int64, Missing}, :hrs / (52*20))) # Create P as defined in class (0,1,2) and create H as a (0,1,2) indicator on hrs worked (0,20,40)
    @subset .!ismissing.(:P) .&& .!ismissing.(:H) # Drop missing values
    @transform @byrow :wage = begin # Create wage variable
        if :hrs>0 && :earn>0 # If hours worked and earnings are positive
            return :earn / :hrs / :CPIU # Calculate wage
        else
            return missing # Else return missing
        end
    end 
end

25615×39 DataFrame
   Row │ Column1  year   MID      intnum  def_flag  age    knum   num_child  a ⋯
       │ Int64    Int64  Int64    Int64?  Bool      Int64  Int64  Int64      F ⋯
───────┼────────────────────────────────────────────────────────────────────────
     1 │       5   1994     4031    5382     false     21      3          2    ⋯
     2 │       6   1995     4031    2239     false     22      3          2
     3 │       7   1996     4031    4246     false     23      3          3
     4 │       9   1998     4031    1702     false     25      3          3
     5 │      11   2000     4031     285     false     27      3          3    ⋯
     6 │      13   2002     4031    1427     false     29      3          3
     7 │      15   2004     4031    1157     false     31      3          3
     8 │      17   2006     4031     196     false     33      3          3
   ⋮   │    ⋮       ⋮       ⋮       ⋮        ⋮        ⋮      ⋮        ⋮        ⋱
 25609 │  137768   1992  9288001    5189     false     24      2          2    ⋯
 25610 │  137815   1991  9292002    9685     false     39      5          3
 25611 │  137816   1992  9292002    7118     false     40      5          3
 25612 │  137884   1991  9293001    9497     false     30      4          4
 25613 │  137885   1992  9293001    4842     false     31      4          4    ⋯
 25614 │  137975   1991  9308002    9496     false     25      2          2
 25615 │  137976   1992  9308002    4839     false     26      2          2
                                               31 columns and 25600 rows omitted

In [2]:
data_mle = (;P = Int64.(data.P), H = Int64.(data.H), year = data.year, age = data.age,
            soi = data.SOI, num_kids = data.num_child, cpi = data.CPIU,
            logwage = log.(coalesce.(data.wage,1.)),wage_missing = ismissing.(data.wage))

(P = [2, 2, 2, 1, 0, 0, 0, 2, 1, 0  …  2, 2, 1, 2, 0, 0, 1, 1, 1, 1], H = [0, 0, 0, 0, 0, 0, 0, 0, 0, 2  …  0, 0, 2, 2, 2, 2, 0, 0, 0, 0], year = [1994, 1995, 1996, 1998, 2000, 2002, 2004, 2006, 2008, 1990  …  1991, 1992, 1991, 1992, 1991, 1992, 1991, 1992, 1991, 1992], age = [21, 22, 23, 25, 27, 29, 31, 33, 35, 17  …  21, 22, 23, 24, 39, 40, 30, 31, 25, 26], soi = [43, 43, 43, 43, 43, 43, 43, 43, 43, 17  …  44, 44, 7, 7, 5, 5, 39, 39, 39, 39], num_kids = [2, 2, 3, 3, 3, 3, 3, 3, 3, 1  …  2, 2, 2, 2, 3, 3, 4, 4, 2, 2], cpi = [0.860812349005761, 0.884959812302546, 0.910948243820851, 0.946664188812488, 1.0, 1.04457233785542, 1.09707768072849, 1.17054218546739, 1.25008130459022, 0.758792510685746  …  0.790785866939231, 0.8148346032336, 0.790785866939231, 0.8148346032336, 0.790785866939231, 0.8148346032336, 0.790785866939231, 0.8148346032336, 0.790785866939231, 0.8148346032336], logwage = [0.0, 0.0, 0.0, 0.0, 1.2039728043259361, 0.0, 0.0, 0.0, 0.0, 0.8230555833449215  …  0.0, 0.0, 2.651257

##### The Model

In class, we defined $P_j \in \{0,1,2\}$ as the indicator for welfare participation, and above we have created $H_j \in \{0,1,2\}$ to indicate the choice over hours worked, with $0$ representing no hours worked, $1$ representing part-time and $2$ representing full-time employment. We are therefore going to consider a static model with 9 discrete choices. Note that there is a Type 1 extreme value shock iid over time and across individuals, $\epsilon_j$ for choice $j$. The shocks are scaled by $\sigma$.

Utility for individual $i$ at time $t$ is given by:
$$
U_{itj} = \log(\max\{50, Y_{it}(W_{it}H_j)\}) + \alpha_l\log(112-H_j)-\alpha_{P,1}1\{P_j>0\} - \alpha_{P,2}1\{P_j>1\}
$$
where $Y_{it}$ is net income for individual $i$ at time $t$ and depends on their earnings $(W_{it}H_j)$, the number of children in the household, and the policy environment in the individual's state of residenceeeee. 


Wages are a deterministic function of age only:
$$
\log(W_{it}) = \gamma_0 + \gamma_1Age_{it}+ \gamma_2Age_{it}^2
$$
and are measured with normally distributed iid measurement error:
$$
\log(W_{it}^0) = \log(W_{it}) + \xi_{it}, \quad  \xi_{it} \sim N(0, \sigma_W^2)
$$

Next, we calculate utility and index choices.

In [3]:
j_idx(p,h) = p*3 + h + 1 # index the agent's choice (p*3 since we have 3 choices for P, then add h, and finally add 1 to start at 1)
function utility(p,h,soi,cpi,year,num_kids,wage,pars::NamedTuple)
    (;αl,Hgrid,σ,αP) = pars
    hrs = Hgrid[1+h]
    earn = wage * hrs
    net_income = max(50.,Transfers.budget(earn,0.,soi,year,num_kids,cpi,p))
    return log(net_income) + αl*log(112-hrs)-αP[1]*(p>0)-αP[2]*(p>1)
end

pars = (;αl = 1., σ=1.,σW=1.,γ= zeros(3),
        Hgrid = [0.,20.,40.], αP = zeros(2))

(αl = 1.0, σ = 1.0, σW = 1.0, γ = [0.0, 0.0, 0.0], Hgrid = [0.0, 20.0, 40.0], αP = [0.0, 0.0])

Choice probabilities are given by the logit formula.

In [4]:
function choice_prob!(logP,it,data,pars) # note that logP is modified in place by the ! in the function name
    (;σ,γ) = pars # unpack
    wage = exp(γ[1] + γ[2] * data.age[it] + γ[3] * data.age[it] ^ 2)
    denom = 0.
    umax = -Inf
    for p in 0:2
        for h in 0:2
            j = j_idx(p,h)
            u = utility(p,h,data.soi[it],data.cpi[it],data.year[it],data.num_kids[it],wage,pars)
            logP[j] = u/σ # log probabilities of each choice
            umax = u>umax ? u : umax # if u > umax, umax = u, else umax = umax
        end
    end
    logP[:] .+= umax/σ # elementwise subtraction #### ???why do we subtract umax/σ rather than add???
    denom = log(sum(exp.(logP))) # log of sum of exponentials
    logP[:] .-= denom # normalise choice probabilities
end

choice_prob! (generic function with 1 method)

Now define the log-likelihood.

In [24]:
function log_likelihood(logP,it,data,pars)
    (;γ,σW) = pars
    logL = 0.
    if !data.wage_missing[it]
        logL += -(0.5(data.logwage[it]-γ[1]-γ[2]*data.age[it]-γ[3]*data.age[it]^2)/σW)^2/2 - log(σW)
    end
    choice_prob!(logP,it,data,pars)
    j = j_idx(data.P[it],data.H[it])
    logL += logP[j]
    return logL
end


function update(x,p)
    αl = exp(x[1])
    σ = exp(x[2])
    γ = x[3:5] # slice x from 3 to 5
    σW = exp(x[6])
    αP = x[7:8]
    updated_pars = (;αl=αl,σ=σ,σW=σW,γ=γ,αP=αP,Hgrid=p.Hgrid)
    return updated_pars
end


function log_likelihood(x,data,pars)
    pars = update(x,pars)
    logP = zeros(eltype(x),9)
    logL = 0.
    for it in eachindex(data.P)
        logL += log_likelihood(logP,it,data,pars)
    end
    return logL
end


log_likelihood (generic function with 2 methods)

Finally, use Optim to maximise the log-likelihood:

In [25]:
x0 = zeros(8)
x0 = [0.,0.,log(10.),0.,0.,0.,0.,0.]
res = optimize(x->-log_likelihood(x,data_mle,pars),x0,BFGS(),autodiff=:forward,Optim.Options(show_trace=true))
pars = update(res.minimizer,pars)

Iter     Function value   Gradient norm 
     0     6.863380e+04     4.644689e+06
 * time: 0.003999948501586914
     1     6.783162e+04     9.075223e+04
 * time: 0.47699999809265137
     2     6.320997e+04     6.805622e+04
 * time: 0.9919998645782471
     3     5.759597e+04     9.535883e+06
 * time: 1.2939999103546143
     4     5.547178e+04     6.980529e+06
 * time: 1.5319998264312744
     5     5.231364e+04     4.581265e+06
 * time: 2.0329999923706055
     6     5.177341e+04     4.205582e+06
 * time: 2.2939999103546143
     7     5.086346e+04     3.405375e+06
 * time: 2.5949997901916504
     8     5.065194e+04     1.745647e+06
 * time: 2.8489999771118164
     9     4.903188e+04     7.185106e+05
 * time: 3.0399999618530273
    10     4.372469e+04     3.577626e+06
 * time: 3.3429999351501465
    11     4.074750e+04     3.265883e+06
 * time: 3.6579999923706055
    12     4.005155e+04     1.697897e+06
 * time: 3.822999954223633
    13     3.979198e+04     8.259920e+05
 * time: 3.96299982

(αl = 2.8546233783929407, σ = 0.9509286759263887, σW = 0.36124300314682983, γ = [-0.02362075058278817, 0.10877398375254806, -0.0012722347394311985], αP = [2.035301636135558, 1.0137619542631473], Hgrid = [0.0, 20.0, 40.0])

##### **Question 1**

**Write code to calculate the standard errors for the parameters and present the estimates in a table.**

Recall that the variance of an ML estimator, $\hat{\theta}^{ML}$ is calculated by the inverse of the Information matrix:
$$
var(\theta) = [I(\theta)]^{-1}
$$
and the Information matrix is the negative of the expected value of the Hessian matrix:
$$
[I(\theta)] = -E[H(\theta)]
$$
and we know that the Hessian is the matrix of the second derivatives of the likelihood with respect to the parameters:
$$
H(\theta) = \frac{\partial \log(L(\theta))}{\partial\theta\partial\theta'}.
$$
By Cramer-Rao, the variance of any unbiased estimator of a parameter $\theta$ must satisfy
$$
var(\theta) \geq (-E[H(\theta)])^{-1}
$$
and then we have that, 
$$
var(\hat{\theta}) = \left(-E\left[\frac{-\partial^2 L}{\partial\theta\partial\theta'}\right]\right)^{-1}
$$
and so, the standard errors of $\hat{\theta}$ are the square roots of the diagonal elements of the Hessian.
Hence, let us use the ForwardDiff package to find the Hessian at the solution to the maximum likelihood estimation, invert the negative of said Hessian, and then find the square roots of the diagonal elements of said matrix. 

In [8]:
using ForwardDiff, LinearAlgebra

optimal_solution = res.minimizer # Get the optimal solution

hessian = ForwardDiff.hessian(x -> log_likelihood(x, data_mle, pars), optimal_solution) # Compute the Hessian at the optimal solution

inv_hessian = inv(-hessian) # Calculate the inverse of the negative of the Hessian (Cramer-Rao lower bound)

std_errors = sqrt.(diag(inv_hessian)) # The standard errors are the square root of the diagonal elements of the above

8-element Vector{Float64}:
 0.012843164746411556
 0.019154147770820935
 0.07099074793657956
 0.004225945782349037
 6.071678210751469e-5
 0.005405767081368025
 0.024051654161771664
 0.022764191318278566

Recall that the parameters we estimated were $\{\alpha_l, \sigma, \sigma_W, \gamma_0, \gamma_1, \gamma_2, \alpha_{P,1}, \alpha_{P,2}\}$  The table below summarises the estimates and their standard errors. Note that I present the information to 5 decimal places.

| Parameter     | Estimate  | Standard Error |
|:-------------:|:---------------:|:-------------:|
| $\alpha_l$        | 2.85462          | 0.01284      |
| $\sigma $        | 0.95093         | 0.01915        |
| $\sigma_W$        | 1.02175         | 0.07099     |
| $\gamma_0 $        | -0.02362          | 0.00422      |
| $\gamma_1$        | 0.10877         | 0.00001        |
| $\gamma_2 $        | -0.00127         | 0.00541      |
| $\alpha_{P,1} $        | 2.03530          |  0.02405       |
| $\alpha_{P,2} $        | 1.01376          |  0.02276     |


#### **Question 2**

**What sources of variation are identifying $\sigma$, the responsiveness of choices to changes in the payoffs? How is the responsiveness of programme participation and labour supply to changes in incentives limited in this framework?**

#### **Question 3**

**Re-estimate the model with a one layer *nested logit* structure, where the individual first makes one of the three participation choices before making one of the three hours choices. That is, if the choices can be ordered as:**
$$
\{P_j\}_{j=1}^9 = \{0,0,0,1,1,1,2,2,2\}, \quad \{H_j\}_{j=1}^9 = \{0,20,40,0,20,40,0,20,40\}
$$
then the partition is:
$$
B = \{\{1,2,3\},\{4,5,6\},\{7,8,9\} \}
$$
**Let $\sigma_H$ be the dispersion of shocks within each partition and $\sigma_P$ be the dispersion of shocks across partitions. Intuitively, how are these dispersion parameters identified?**

In this question, we need to re-estimate the model with the *nested logit* structure. Relative to what we saw in class, where individuals chose whether to work $0$ hours or $30$ hours in order to meet the work requirements, here agents choose whether to not work, work part-time, or work full-time. Denoting the dispersion of shocks *within* each partition as $\sigma_H$ and the dispersion of shocks *across* partitions as $\sigma_P$, we therefore see that we need to estimate $\{\alpha_l, \sigma_H, \sigma_P, \sigma_W, \gamma_0, \gamma_1, \gamma_2, \alpha_{P,1}, \alpha_{P,2}\}$.

In [9]:
function utility_nest(p,h,soi,cpi,year,num_kids,wage,params::NamedTuple)
    (;αl,Hgrid_nested,αP) = params
    hrs = Hgrid_nested[1+h]
    earn = wage * hrs
    net_income = max(50.,Transfers.budget(earn,0.,soi,year,num_kids,cpi,p))
    return log(net_income) + αl*log(112-hrs)-αP[1]*(p>0)-αP[2]*(p>1)
end

utility_nest (generic function with 1 method)

In [20]:
params = (;αl = 1., σH = 1., σP = 1.,σW=1.,γ= zeros(3),
        Hgrid_nested = [0.,20.,40.], Pgrid_nested = [0.,1.,2.],
        αP = zeros(2))
 
function choice_probs_nest!(logP_nest, it, data, params)
    (;σH,σP,γ) = params
    wage = exp(γ[1] + γ[2] * data.age[it] + γ[3] * data.age[it] ^ 2)
    u = zeros(eltype(logP_nest), 9)
    for p in 0:2
        for h in 0:2
            j = j_idx(p,h)
            u[j] = utility_nest(p,h,data.soi[it],data.cpi[it],data.year[it],data.num_kids[it],wage,params) #  utility for each alternative
        end
    end
    u_exp = exp.(u / σH)
    IV = σH * log.([sum(u_exp[1:3]), sum(u_exp[4:6]), sum(u_exp[7:9])]) # calculate the IV for each nest ### do we need an extra σH here??

    IV_exp = exp.(IV / σP) # exponentiate the IV and divide by σP
    nest_probs = IV_exp / sum(IV_exp) # probability of choosing a nest

    for nest in 1:3
        u_nest = log.(u_exp[(nest-1)*3+1:nest*3]) # extract the utility for the nest by slicing the utility vector
        denom = log.(sum(u_nest)) # log of the sum of the exponentiated utility for the nest
        u_nest[:] .-= denom # subtract the above from the utility for the nest
        logP_nest[(nest-1)*3+1:nest*3] = log.(nest_probs[nest]) .+ u_nest # the log choice prob for a given alternative is the log of the nest prob plus the log of the within nest prob
    end
end

choice_probs_nest! (generic function with 1 method)

In [32]:
function log_likelihood_nest(logP_nest, it, data, params)
    (;γ,σW) = params
    logL_nest = 0.
    if !data.wage_missing[it]
        logL_nest += -(0.5(data.logwage[it]-γ[1]-γ[2]*data.age[it]-γ[3]*data.age[it]^2)/σW)^2 - log(σW)
    end
    choice_probs_nest!(logP_nest, it, data, params)
    j = j_idx(data.P[it], data.H[it])
    logL_nest += logP_nest[j]
    return logL_nest
end

function update_nest(x, params)
    αl = exp(x[1])
    σH = exp(x[2])
    σP = exp(x[3])
    σW = exp(x[4])
    γ = x[5:7]
    αP = x[8:9]
    Hgrid_nested = [0.,20.,40.]
    Pgrid_nested = [0.,1.,2.]
    return (;params..., αl, σH, σP, σW, γ, αP, Hgrid_nested, Pgrid_nested)
end

function log_likelihood_nest(x, data, params)
    params = update_nest(x, params)
    logP_nest = zeros(eltype(x), 9)
    logL_nest = 0.
    for it in eachindex(data.P)
        logL_nest += log_likelihood_nest(logP_nest, it, data, params)
    end
    return -logL_nest
end

log_likelihood_nest (generic function with 2 methods)

In [35]:
x0_nest = zeros(9)
x0_nest = [0.11,0.11,0.3,0.11,0.11,0.11,0.11,0.11,0.11]
res_nest = optimize(x->-log_likelihood_nest(x,data_mle,params),x0_nest,BFGS(),autodiff=:forward,Optim.Options(show_trace=true))
params = update_nest(res_nest.minimizer,params)

Iter     Function value   Gradient norm 
     0    -6.693593e+07     1.219414e+09
 * time: 0.0
     1    -1.162874e+08     1.612739e+09
 * time: 14.450999975204468
     2    -1.162874e+08     1.612739e+09
 * time: 26.44700002670288


(αl = 1.1162745115513852, σH = 1.1163407667239995, σP = 1.3498589297937567, σW = 1.1119633105990736, γ = [0.11002430007428211, 0.11088909151168083, 0.14433021079038777], Hgrid_nested = [0.0, 20.0, 40.0], Pgrid_nested = [0.0, 1.0, 2.0], αP = [0.11000006745110112, 0.1099999832639781])

#### **Question 4**

**What sort of variation do you think is in the data that we are failing to put in our model? What do you think are the implication for key parameter estimates?**